### ISSUE TITLE
[no-confusing-void-expression] Add an option to warn on non void arrow callbacks

### ISSUE THREAD

**Issue by ArnaudBarre** [typescript-eslint/typescript-eslint#2988]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-confusing-void-expression": [
      "warn",
      { ignoreArrowShorthand: true },
    ],
  }
}
```

```TS
// setState return void, this is ok and shorter
<Modal onClose={() => setState(undefined)} /> 
// We are returning a number to a void callback. This is probably a mistake
<Counter onChange={(value) => Math.abs(value)}   /> 
```

**Expected Result**

Warn in the second case. 

I would like the rule to report on return type mismatch like `no-misused-promises`, while keeping the the possibility to use arrow function when callback returns nothing. Would an option for this be accepted? I can work on it next week if I have a green light

We could also change the default behaviour of ignoreArrowShorthand, but it would be a breaking change

**Actual Result**

Warn in both cases by default, non with `{ ignoreArrowShorthand: true }`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.1` |
| `@typescript-eslint/parser`        | `4.14.1` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.19.0` |
| `node`                             | `4.13.1` |

**Comment by bradzacher** [comment]
I don't think this belongs in the `no-confusing-void-expression` rule.
That rule is about passing/returning a void type.

There is definitely space for a new rule that does this form of check, however.

**Comment by phaux** [comment]
I'm gonna send a PR for this one soon:tm:

**Comment by theorlovsky** [comment]
any updates on this one?

**Comment by bradzacher** [comment]
With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.

We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.

If not - please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam ***everyone*** subscribed to the issue.

**Comment by phaux** [comment]
@theorlovsky 
> any updates on this one?

I started working on this 3 months ago but haven't finished it yet. It wasn't as easy as I thought. I'm gonna finish it eventually tho
